### PR TITLE
Allow subscription to indicate that a userlimit is reached

### PR DESCRIPTION
**Pull request by MorrisJobke**

### CODE REVIEW

File: lib/private/Support/Subscription/Registry.php [hunk 1]
```diff
@@ -127,9 +138,67 @@ public function delegateHasValidSubscription(): bool {
...
+
+		$disabledUsers = $this->config->getUsersForUserValue('core', 'enabled', 'false');
+		$disabledUsersCount = count($disabledUsers);
+		$this->userCount = $userCount - $disabledUsersCount;
+
+		if ($this->userCount < 0) {
+			$this->userCount = 0;
+
+			// TODO this should never happen
+			$this->logger->warning("Total user count was negative (users: $userCount, disabled: $disabledUsersCount)", ['app' => 'support']);
```
Comment (rullzer) [inline lib/private/Support/Subscription/Registry.php:199]: famous last words

File: lib/private/Support/Subscription/Registry.php [hunk 2]
```diff
@@ -127,9 +138,67 @@ public function delegateHasValidSubscription(): bool {
...
+				} else {
+					// TODO what if the user count can't be determined?
+					$this->logger->warning('Can not determine user count for ' . get_class($backend), ['app' => 'support']);
+				}
+			}
+		}
+
+		$disabledUsers = $this->config->getUsersForUserValue('core', 'enabled', 'false');
+		$disabledUsersCount = count($disabledUsers);
+		$this->userCount = $userCount - $disabledUsersCount;
```
Comment (rullzer) [inline lib/private/Support/Subscription/Registry.php:193]: why do we store this? It should oly be called once right?

**Review by rullzer — CHANGES_REQUESTED** [review]
Nice stuff!

File: lib/private/Support/Subscription/Registry.php (hunk 2 above)
Comment (MorrisJobke) [reply lib/private/Support/Subscription/Registry.php:193]: copy paste leftover :/

File: lib/private/Support/Subscription/Registry.php [hunk 3]
```diff
@@ -127,9 +139,76 @@ public function delegateHasValidSubscription(): bool {
...
+
+			// this should never happen
+			$this->logger->warning("Total user count was negative (users: $userCount, disabled: $disabledUsersCount)", ['app' => 'lib']);
+		}
+
+		return $userCount;
+	}
+
+	private function notifyAboutReachedUserLimit() {
+		// TODO notify admin about reached user limit
```
Comment (MorrisJobke) [inline lib/private/Support/Subscription/Registry.php:211]: @rullzer How do we want to do this? Email? Notification only?

Comment (MorrisJobke) [reply lib/private/Support/Subscription/Registry.php]: Implemented as notification via 2260c02e97

**Comment by MorrisJobke** [comment]
Ready for review 🚀

**Review by rullzer — APPROVED** [review]
Looks good! Didn't test in depth but a quick smoke test showed it OK. And the test seems sane!

**Review by ChristophWurst — APPROVED** [review]
Looks good!

File: lib/private/User/Manager.php [hunk 4]
```diff
@@ -297,6 +299,12 @@ public function searchDisplayName($pattern, $limit = null, $offset = null) {
 	 * @return bool|IUser the created user or false
 	 */
 	public function createUser($uid, $password) {
+		// DI injection is not used here as IRegistry needs the user manager itself for user count and thus it would create a cyclic dependency
+		if (\OC::$server->get(IRegistry::class)->delegateIsHardUserLimitReached()) {
+			$l = \OC::$server->getL10N('lib');
+			throw new HintException($l->t('The user limit has been reached and the user was not created.'));
+		}
+
```
Comment (blizzz) [inline lib/private/User/Manager.php:307]: heads up, this will not catch users provisioned by other backends

Comment (rullzer) [reply lib/private/User/Manager.php:307]: Yes we know
will come in follup PRs

**Review by jospoortvliet — COMMENTED** [review]
Looks good, just these changes would be nice :+1:

File: core/Notification/CoreNotifier.php [hunk 5]
```diff
@@ -73,6 +74,13 @@ public function prepare(INotification $notification, string $languageCode): INot
 			return $notification;
 		}
 
+		if ($notification->getSubject() === 'user_limit_reached') {
+			$notification->setParsedSubject($l->t('The user limit of this instance is reached.'));
+			$notification->setParsedMessage($l->t('Add a subscription key to increase the user limit of this instance. For more information have a look at the Enterprise subscription page.'));
+			$notification->setLink('https://nextcloud.com/enterprise/order/');
```
[the remark this answers is not in the document]
Comment (MorrisJobke) [reply core/Notification/CoreNotifier.php:80]: Added to https://github.com/nextcloud/server/pull/24516

File: core/Notification/CoreNotifier.php [hunk 6]
```diff
@@ -73,6 +74,13 @@ public function prepare(INotification $notification, string $languageCode): INot
 			return $notification;
 		}
 
+		if ($notification->getSubject() === 'user_limit_reached') {
+			$notification->setParsedSubject($l->t('The user limit of this instance is reached.'));
+			$notification->setParsedMessage($l->t('Add a subscription key to increase the user limit of this instance. For more information have a look at the Enterprise subscription page.'));
```
[the remark this answers is not in the document]
Comment (MorrisJobke) [reply core/Notification/CoreNotifier.php:79]: Added to https://github.com/nextcloud/server/pull/24516

**Pull request merged**